### PR TITLE
[clang] When checking for covariant return types, make sure the pointers or references are to *classes*

### LINKED ISSUE #111742: Clang wrongly allows overriding method to vary its return type
The following code generates no warnings in Clang despite the override's return type being incorrect:
```
struct S {
  virtual const void* f() = 0;
  virtual ~S() = default;  // Silence -Wnon-virtual-dtor
};

struct T : S {
  void* f() override;  // Should warn or error
};
```
Other compilers:
gcc: Warns
MSVC: Errors
Live example: https://godbolt.org/z/TjoxEzd7M

Larger example demonstrating dispatch in the presence of overloads: https://godbolt.org/z/orcfbrnGz

### PR DESCRIPTION
**Pull request by bricknerb**
https://eel.is/c++draft/class.virtual#8.1

This prevents overriding methods with non class return types that have less cv-qualification.

Fixes: #111742

### CODE REVIEW

**Review by Sirraide — COMMENTED** [review]
Implementation seems fine, but this is still missing a release note (in `clang/docs/ReleaseNotes.rst`, probably somewhere in the ‘Potentially Breaking Changes’ section).

CC @AaronBallman @cor3ntin Do we want to make this a warning that defaults to an error instead? Imo it’s probably fine as-is because I don’t imagine too many people would write code that abuses us not diagnosing this...

**Comment by cor3ntin** [comment]
>  CC @AaronBallman @cor3ntin Do we want to make this a warning that defaults to an error instead? Imo it’s probably fine as-is because I don’t imagine too many people would write code that abuses us not diagnosing this...

It looks fine to me as is.

Code that would do something like `static_cast<BaseReturnType>(derived->f())` (which I think is perfectly reasonable)  would explode, despite the standard guarantees it should work.

If people scream we have plenty of time to reconsider making it a warning

File: clang/test/SemaCXX/virtual-override.cpp [hunk 1]
```diff
@@ -289,3 +289,13 @@ namespace PR8168 {
...
+
+namespace T13 {
+  struct A {
+    virtual const int *f() const; // expected-note{{overridden virtual function is here}}
+  };
+
+  struct B : A {
+    int *f() const override; // expected-error{{virtual function 'f' has a different return type ('int *') than the function it overrides (which has return type 'const int *')}}
+  };
+}
```
Comment (cor3ntin) [inline clang/test/SemaCXX/virtual-override.cpp:301]: can we have additional tests with various combination of `void*`, `int*` and some class type?

Comment (bricknerb) [reply clang/test/SemaCXX/virtual-override.cpp:301]: Do you mean that base return type A (`void*`, `int*` or some class type) and derived returns type B != A (`void*`, `int*` or some class type)?
Or do you mean they both return the same type but it's qualified differently?

For different types we have T1 and T2.
For different qualification of class types, we have T6.
Let me know what use case you think is uncovered.

Comment (ilya-biryukov) [reply clang/test/SemaCXX/virtual-override.cpp:301]: Given that the code has a different treatment of types which are pointers and references, having a variation of `T2` with non-class types seems valuable. (we could probably add it to `T2` directly).

Same for qualification in `T6`, given that the rules are special for the class types.

Comment (bricknerb) [reply clang/test/SemaCXX/virtual-override.cpp]: Added tests in T2 and T6, and removed T13.
Please let me know if this is not what you meant.

**Review by cor3ntin — APPROVED** [review]
LGTM

**Review by ilya-biryukov — APPROVED** [review]
LGTM

**Pull request merged**